### PR TITLE
Add more diagnostic channels

### PR DESCRIPTION
**Pull request by fatal10110**
Trying to close https://github.com/nodejs/undici/issues/1032

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
Nice work! Could you add tests and docs?

**Review by mcollina — COMMENTED** [review]
I think we should pass the socket to each event.

**Comment by fatal10110** [comment]
> I think we should pass the socket to each event.

The only channels that can pass the socket is "connected" and "write"
should I add it?

**Comment by mcollina** [comment]
I would say so. Do we need the write publish?

**Comment by mcollina** [comment]
You might just add the client everywhere.

**Comment by fatal10110** [comment]
> Do we need the write publish?

I think yes, as discussed here, but maybe i'm wrong
https://github.com/nodejs/undici/issues/1032#issuecomment-920816090

> You might just add the client everywhere.
@ronag does not want to "have a dependency on the client object."
https://github.com/nodejs/undici/issues/1032#issuecomment-918901681

**Comment by fatal10110** [comment]
I've passed the socket, to "write" and "connect" channels. I can pass it to all channels except "create" it will require client to pass the socket to some callbacks, but sometimes the socket may be already destroyed or missing (on "error")

**Review by ronag — APPROVED** [review]

**Comment by mcollina** [comment]
Can you run some benchmarks?

**Review by mcollina — APPROVED** [review]
lgtm

File: docs/api/DiagnosticsChannel.md
Comment (szmarczak) [inline docs/api/DiagnosticsChannel.md:83]: `write` -> `sendHeaders`

File: docs/api/DiagnosticsChannel.md
Comment (szmarczak) [inline docs/api/DiagnosticsChannel.md:98]: `connect` -> `beforeConnect`

**Review by szmarczak — APPROVED** [review]

**Comment by artur-ma** [comment]
This branch without subscribers

image

This branch with empty subscribers

image

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/client.js
Comment (bizob2828) [inline lib/client.js:1464]: this def will achieve what i mentioned in https://github.com/nodejs/undici/issues/1032 however the headers are a string so it's now on the user to parse that string which to me is not ideal. Is there a reason why the headers aren't associated with the request all the time?

Comment (bizob2828) [reply lib/client.js:1464]: I guess you can also get what you need from the socket object so my previous comment/question may be moot

File: lib/client.js
Comment (bizob2828) [inline lib/client.js:1117]: you may need to attach the client.  i tried subscribing but I had no object to attach some context in the connected and/or connectError to properly end spans

File: lib/client.js
Comment (fatal10110) [reply lib/client.js:1464]: > I guess you can also get what you need from the socket object 

Not sure I understood what you are trying to get form the socket

> so it's now on the user to parse that string 

Undici operates headers as string, IMO the parsing should not be done by undici this extra work may affect performance, but the user may not use it at all, the headers string follows the HTTP standard (\r\n separator) should be easy to parse

Comment (bizob2828) [reply lib/client.js:1464]: In #1032 I requested outgoing headers, specifically the host header.  Since headers in client are just a string and not key/values pairs like in request it would've been a pain to parse a string to extract the host header.  However, since the socket is here it has the host that it is connected to which is the same so I can just rely on the socket object to get what I need instead of parsing headers string

File: lib/client.js
Comment (bizob2828) [reply lib/client.js:1117]: just to be clear, the client will need to be on all 3 connect events

File: lib/client.js
Comment (fatal10110) [reply lib/client.js:1464]: > key/values pairs like in request

If u add any headers to request, please note, on request object they are also string a string 
and should be added as follow `request.addHeader('hello', 'world')`

File: lib/client.js
Comment (ronag) [reply lib/client.js:1117]: Probably more appropriate to attach he connector?

File: lib/client.js
Comment (bizob2828) [reply lib/client.js]: yep, you can see in this prototype I'm using `addHeader` https://github.com/newrelic/node-newrelic/pull/901/files#diff-9fe41f6ad1e32c80844e1948f016e7d692ea907edd2620e8b9245c1c9da67b37R70

File: lib/client.js
Comment (bizob2828) [reply lib/client.js:1117]: looks like the connector is a function, if so then I would only need to subscribe to beforeConnect and properly bind the async function

Comment (fatal10110) [reply lib/client.js:1117]: The only benefit from passing the client, is that both `connect` and `sendHeaders` will have the same context
If we will pass connector instead of client then I would not be able to send it to `sendHeaders` 

@bizob2828 what do u prefer?

File: lib/client.js
Comment (fatal10110) [reply lib/client.js]: Here u are using `request.headers` as object, but it is a string 

```
https://github.com/newrelic/node-newrelic/pull/901/files#diff-9fe41f6ad1e32c80844e1948f016e7d692ea907edd2620e8b9245c1c9da67b37R54
```

see this test
https://github.com/nodejs/undici/blob/main/test/diagnostics-channel/get.js#L41

File: lib/client.js
Comment (bizob2828) [inline lib/client.js:1125]: `this` is not the client in this context

File: lib/client.js
Comment (bizob2828) [reply lib/client.js:1117]: I don't need the same context between sendHeaders and connect, But I'm fine with passing the client.  I'm very new to this lib and I just built instrumentation. @ronag appears to be involved more heavily so I will defer to him

File: lib/client.js
Comment (bizob2828) [reply lib/client.js]: yea i noticed that. Honestly this is copied code from our http-outbound and I don't see how a request header could ever be a symbol when serialized

File: lib/client.js
Comment (bizob2828) [reply lib/client.js:1117]: If we pass the connector in beforeConnect I don't see any value add with adding the connector in the connected and connectError. Typically you'd be binding the function in the beforeConnect and the other two events will be handled in the instrumentation primitives of the given vendor

**Comment by mcollina** [comment]
wdyt @ronag?

File: lib/client.js
Comment (bizob2828) [reply lib/client.js:1117]: > If we pass the connector in beforeConnect I don't see any value add with adding the connector in the connected and connectError. Typically you'd be binding the function in the beforeConnect and the other two events will be handled in the instrumentation primitives of the given vendor

I see the maintainers have approved. don't sweat this. it'll still work and depending on vendor methods of attaching context this will work just fine

**Comment by bizob2828** [comment]
@ronag @mcollina I'm about to open a PR for the New Relic agent that adds support for undici.  I will need this to be released so I was wondering when the next undici release was going to be published to npm?

**Comment by mcollina** [comment]
We usually release once a week or similar.

**Comment by bizob2828** [comment]
> We usually release once a week or similar.

Great, I'll keep an eye out.

**Pull request merged**
